bee-common/src/shutdown/mod.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+use crate::worker::Error as WorkerError;
+
+use async_std::task;
+use futures::channel::oneshot;
+use thiserror::Error;
+
+use std::result::Result;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Sending the shutdown signal to a task failed.")]
+    SendingShutdownSignalFailed,
+
+    #[error("Waiting for task to end failed.")]
+    WaitingforWorkerShutdownFailed(#[from] WorkerError),
+}
+
+pub type ShutdownNotifier = oneshot::Sender<()>;
+pub type ShutdownListener = oneshot::Receiver<()>;
+pub type WorkerHandle = task::JoinHandle<Result<(), WorkerError>>;
+
+/// Handles the graceful shutdown of asynchronous tasks.
+pub struct ShutdownHandler {
+    notifiers: Vec<ShutdownNotifier>,
+    workers: Vec<WorkerHandle>,
+    actions: Vec<Box<dyn Fn()>>,
+}
+
+impl ShutdownHandler {
+    /// Creates a new instance.
+    pub fn new() -> Self {
+        Self {
+            notifiers: vec![],
+            workers: vec![],
+            actions: vec![],
```

Comment:
Maybe only me but I find `Vec::new()` more clear and readable than `vec![]`

Reply:
Or could simply derive Default and use it

Reply:
yea, I am using vec![] since I started with Rust and probably just got used to it. For me personally I find **it** more readable because of the eye catching brackets and the !, but I don't mind so much. We should probably use `Vec::with_capacity` here anyway.

Reply:
Yeah, just find it more logical to find `new` methods while implementing a `new` method.
Hmm what would be the capacity then ?

Reply:
idk....8 or 16, something like that